### ISSUE TITLE
2nd factor authentication turns off 1st factor authentication

### ISSUE THREAD

**Issue by clefru** [NixOS/nixpkgs#14997]
When I did "security.pam.services.sshd.oathAuth = true", the resulting pam file was:

auth sufficient pam_unix.so  likeauth try_first_pass
auth sufficient ${pkgs.oathToolkit}/lib/security/pam_oath.so window=5 usersfile=/etc/users.oath digits=6

which allows me to type a wrong password and make up for it by a correct OTP. I haven't reviewed the resulting PAM file carefully after making that change, and I was taken by surprise by that behavior. I would have expected a more conservative "conjunctive" semantic here, as in "OTP" is required in addition to everything else. 

Justification:
While the other authentication mechanisms surrounding the "auth" directives in 
https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/security/pam.nix#L262
could be argued to have password like qualities and rightfully replace pam_unix.so (pam_ldap, pam_otpw), I would argue that 2nd factor auth methods like xOTP don't have those qualities (2^6^8 for a 8 char password vs 10^6~=2^16 for xOTP). Otherwise, they wouldn't be called 2nd factor :)

Implementation:
Move 2nd factor mechanisms to the top of the auth section and mark them required. Unfortunately that makes all authenticators prompt for 2nd factor first, which is a bit odd. I haven't yet thought about a more complex restructuring of pam.nix.

While at it, we could split up the "pam.text" section into four overridable ones for account/auth/session/password, so that people, that don't agree with my conservative stance and want 2nd factor only, can change their configs easily.
- System: 16.03.616.4ac4b35 (Emu)
- Nix version: nix-env (Nix) 1.11.2
- Nixpkgs version: 16.03.616.4ac4b35

**Comment by spinus** [comment]
There is PR https://github.com/NixOS/nixpkgs/pull/11886 which tries to fix it...

**Comment by spinus** [comment]
I assume, I could add something like ``security.pam.u2f.type = required|sufficient``, is that right?

**Comment by spacekitteh** [comment]
jesus, does this still occur?

**Comment by jpierre03** [comment]
@clefru is there some updates ?
cc @vcunat @obadz @spacefrogg

**Comment by clefru** [comment]
@jpierre03 Yes.[*] https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/security/pam.nix#L271 still has `auth sufficient pam_unix.so` followed by L275 which generates also `auth sufficient pam_oath.so`. Why this is problematic is explained in the initial comment.

My fix hard-codes:
```
# Authentication management.
auth required pam_unix.so  likeauth
auth required ${pkgs.oathToolkit}/lib/security/pam_oath.so window=5 usersfile=/etc/users.oath digits=6
```
 
[*] Not tested, but I see no changes in pam.nix since I filed this that could affect the pam file generation.

**Comment by spinus** [comment]
I have a PR https://github.com/NixOS/nixpkgs/pull/11886 where I refactored it a little (but u2f not oath), I need to polish it a little and I'll try to fix this issue on u2f module.

**Comment by wizeman** [comment]
For those following this at home, I have submitted a tentative fix for this issue in #22393, which seems to work for me.

It would be great if someone who understands PAM could review it, since I'm far from being an expert.

**Comment by grahamc** [comment]
Fixed in master ^
